### ISSUE TITLE
Prevent initial invocation?

### ISSUE THREAD

**Issue by slavafomin** [node-schedule/node-schedule#236]
Hello!

Thank you for this great module!

However, it looks like `scheduleJob()` invokes specified function initially when called, which is counter-intuitive and extremely undesirable.

How do I prevent initial invocation of the specified function and run it **only** on the specified time?

Thank you!

**Comment by slavafomin** [comment]
The weird thing is that this behavior is not consistent.

Consider this example:

```javascript
      schedule.scheduleJob({
        hour: 7,
        minute: 0
      }, function () {
        sails.log.info('RUNNING JOB');
      });

      schedule.scheduleJob('0 7 * * *', function () {
        sails.log.info('RUNNING JOB 2');
      });
```

`RUNNING JOB 2` is logged as soon as `scheduleJob()` is called. However, `RUNNING JOB` is called only at the specified time.

**Comment by tejasmanohar** [comment]
weird, can replicate w/ certain cron inputs.

**Comment by pgkehle** [comment]
I'm seeing similar behavior, and the schedule does not repeat, when using cron job styling.

**Comment by santigimeno** [comment]
@slavafomin I've tried this code using `sinon` to fake the `Date` and can't reproduce the error:
```
var schedule = require('./');
var sinon = require('sinon');

var now = Date.now();

var clock = sinon.useFakeTimers();

// Set the current date to now
clock.tick(now);

schedule.scheduleJob({
  hour: 7,
  minute: 0
}, function () {
  console.log('RUNNING JOB - ' + new Date());
});

schedule.scheduleJob('0 7 * * *', function () {
  console.log('RUNNING JOB 2 - ' + new Date());
});

// After 10 days...
clock.tick(10 * 24 * 60 * 60 * 1000);
```
outputs:
```
RUNNING JOB - Tue Feb 02 2016 07:00:00 GMT+0100 (CET)
RUNNING JOB 2 - Tue Feb 02 2016 07:00:00 GMT+0100 (CET)
RUNNING JOB - Wed Feb 03 2016 07:00:00 GMT+0100 (CET)
RUNNING JOB 2 - Wed Feb 03 2016 07:00:00 GMT+0100 (CET)
RUNNING JOB - Thu Feb 04 2016 07:00:00 GMT+0100 (CET)
RUNNING JOB 2 - Thu Feb 04 2016 07:00:00 GMT+0100 (CET)
RUNNING JOB - Fri Feb 05 2016 07:00:00 GMT+0100 (CET)
RUNNING JOB 2 - Fri Feb 05 2016 07:00:00 GMT+0100 (CET)
RUNNING JOB - Sat Feb 06 2016 07:00:00 GMT+0100 (CET)
RUNNING JOB 2 - Sat Feb 06 2016 07:00:00 GMT+0100 (CET)
RUNNING JOB - Sun Feb 07 2016 07:00:00 GMT+0100 (CET)
RUNNING JOB 2 - Sun Feb 07 2016 07:00:00 GMT+0100 (CET)
RUNNING JOB - Mon Feb 08 2016 07:00:00 GMT+0100 (CET)
RUNNING JOB 2 - Mon Feb 08 2016 07:00:00 GMT+0100 (CET)
RUNNING JOB - Tue Feb 09 2016 07:00:00 GMT+0100 (CET)
RUNNING JOB 2 - Tue Feb 09 2016 07:00:00 GMT+0100 (CET)
RUNNING JOB - Wed Feb 10 2016 07:00:00 GMT+0100 (CET)
RUNNING JOB 2 - Wed Feb 10 2016 07:00:00 GMT+0100 (CET)
RUNNING JOB - Thu Feb 11 2016 07:00:00 GMT+0100 (CET)
RUNNING JOB 2 - Thu Feb 11 2016 07:00:00 GMT+0100 (CET)
```
Could you create a reduce test so I can debug it?
Thanks

**Comment by santigimeno** [comment]
@pgkehle Any test sample I can use?

**Comment by pgkehle** [comment]
I had this code:

```
  jobs.jRulesTest = nodeScheduler.scheduleJob('*/10 * * * * *', function () {
    log.info('rulesTest()...');
    schedules.rulesTest(function () {
      log.info('rulesTest() exited...');
    });
  });

```

schedules.rulesTest has this: 

```
var locked = false;
schedules.rulesTest = module.exports = function (callback) {
  log.warn('called');

  if (locked === true) {
    log.warn('locked...');
    return callback();
  }

  log.warn('not locked...');
  locked = true;

  "do a bunch of stuff talking to mongoose/mongo, then eventually:"

        log.error("unlocking... ");
        locked = false;
        callback();
};
```

The output ends up with this:

worker-1 info: [schedules/watch.js] called; starting monitoring processes...
worker-1 info: [schedules/watch.js] jRulesTest()...
worker-1 warn: [schedules/rulesTest.js] not locked...
worker-1 warn: [schedules/taskBackgroundProcess.js] called
worker-1 info: [schedules/watch.js] jRulesTest()...
worker-1 warn: [schedules/rulesTest.js] locked...
worker-1 info: [schedules/watch.js] jRulesTest() exited...
worker-1 error: [schedules/rulesTest.js] unlocking...
worker-1 info: [schedules/watch.js] jRulesTest() exited...

This all happens immediately when the PM2 process is restarted. 

Sorry there is no timing in the debug.

**Comment by pgkehle** [comment]
I just wrote this to get by this issue:


```
  var ruleCfgs = [];
  ruleCfgs.push({name: 'rulesTest', timing: [{second: 0}, {second: 10}, {second: 20}, {second: 30}, {second: 40}, {second: 50}]});
  ruleCfgs.push({name: 'taskBackgroundProcess', timing: [{second: 5}, {second: 15}, {second: 25}, {second: 35}, {second: 45}, {second: 55}]});

  _.each(ruleCfgs, function (ruleCfg) {
    log.info('ruleCfg.timing: ' + JSON.stringify(ruleCfg.timing, null, 4));
    _.each(ruleCfg.timing, function (t) {

      log.info(ruleCfg.name + ' t: ' + JSON.stringify(t, null, 4));
      var rr = new nodeScheduler.RecurrenceRule();

      _.each(t, function (val, type) {
        rr[type] = val;
      });

      log.info(ruleCfg.name + ' rr: ' + JSON.stringify(rr, null, 4));

      if (!jobs[ruleCfg.name]) jobs[ruleCfg.name] = [];
      var j = nodeScheduler.scheduleJob(rr, function () {
        log.info('calling ' + ruleCfg.name + '()...');
        schedules[ruleCfg.name](function () {
          log.info('exited ' + ruleCfg.name + '()...');
        });
      });

      jobs[ruleCfg.name].push(j);
    })
  });

```

**Comment by santigimeno** [comment]
@pgkehle What exactly do you want to achieve with that code? Run a task every 5 seconds?

**Comment by santigimeno** [comment]
Ok, I think I understand now. What about using cron style scheduling. The rules would be:
```
rulesTest = '*/10 * * * * *';
taskBackgroundProcess = '5,15,25,35,45,55 * * * * *';
```

**Comment by santigimeno** [comment]
Regarding the issue now. In the initial code you posted you said something about  PM2. Does the problem happens if you run the code by hand?

**Comment by santigimeno** [comment]
@pgkehle I've run your second code and seems to be running just fine. I don't know what the problem might be :(

**Comment by pgkehle** [comment]
@santigimeno Yes, that was the intent, to run something quick periodically, approximately 1x/10 seconds. 

Initially I had 
* rulesTest: '*/10 * * * * *' 
* taskBackgroundProcess: '*/5 * * * * *'

But what the rewrite I happened to just alternate those two. 

The cron-type scheduling worked great int he past.  Now, I'll get the initial firing, then nothing.. no repeating..  So my re-write works for now, using the repeating that IS working using RecurrenceRule().

Regarding PM2, issue is this worked before, no change in code surrounding scheduling.  No, I've not tried running it by hand.

**Comment by pgkehle** [comment]
@santigimeno Should this work?

```
var schedule = require('node-schedule');
var rule = new schedule.RecurrenceRule();

rule.second = "5,15,25,35,45,55";  // this?
rule.second = [5,15,25,35,45,55]; // or this?  
 
var j = schedule.scheduleJob(rule, function(){
  console.log('The answer to life, the universe, and everything!');
});
```

**Comment by santigimeno** [comment]
TBH I don't know. I just use the `cron` syntax. Have you tried?

**Comment by santigimeno** [comment]
BTW, this is your code, slightly modified so I could run it:
```
var nodeScheduler = require('./');
var _ = require('underscore');

var jobs = {};
var log = {
   info : console.log
};

var schedules = {
   rulesTest : function(cb) {
     cb();
   },
   taskBackgroundProcess : function(cb) {
     cb();
   }
};

var watch = module.exports = function (callback) {
  log.info('called; starting monitoring processes...');
  // Until this gets fixed through node-schedule 
  var ruleCfgs = [];
  ruleCfgs.push({name: 'rulesTest', timing: [{second: 0}, {second: 10}, {second: 20}, {second: 30}, {second: 40}, {second: 50}]});
  ruleCfgs.push({name: 'taskBackgroundProcess', timing: [{second: 5}, {second: 15}, {second: 25}, {second: 35}, {second: 45}, {second: 55}]});

  _.each(ruleCfgs, function (ruleCfg) {
    log.info('ruleCfg.timing: ' + JSON.stringify(ruleCfg.timing, null, 4));
    _.each(ruleCfg.timing, function (t) {

      log.info(ruleCfg.name + ' t: ' + JSON.stringify(t, null, 4));
      var rr = new nodeScheduler.RecurrenceRule();

      _.each(t, function (val, type) {
        rr[type] = val;
      });

      log.info(ruleCfg.name + ' rr: ' + JSON.stringify(rr, null, 4));

      if (!jobs[ruleCfg.name]) jobs[ruleCfg.name] = [];
      var j = nodeScheduler.scheduleJob(rr, function () {
        log.info('calling ' + ruleCfg.name + '()...' + new Date());
        schedules[ruleCfg.name](function () {
          log.info('exited ' + ruleCfg.name + '()...' +  new Date());
        });
      });

      jobs[ruleCfg.name].push(j);
    })
  });
};

watch();
```

**Comment by pgkehle** [comment]
Also, FWIW: with the original code of this:

```
    jobs.rulesTest = nodeScheduler.scheduleJob('*/10 * * * * *', function () {
      log.info('calling rulesTest() @ ' + moment().toISOString());
      schedules.rulesTest(function () {
        log.info('rulesTest() exited...');
      });
    });

    jobs.taskBackgroundProcess = nodeScheduler.scheduleJob('*/5 * * * * *', function () {
      log.info('calling taskBackgroundProcess() @ ' + moment().toISOString());
      schedules.taskBackgroundProcess(function (err) {
        log.info('taskBackgroundProcess(), err: ' + err);
      });
    });

```

this is the output:

```
worker-1 info: [schedules/watch.js] called; starting monitoring processes...
worker-1 info: [schedules/watch.js] calling rulesTest() @ 2016-02-01T16:02:52.137Z
worker-1 info: [schedules/watch.js] calling taskBackgroundProcess() @ 2016-02-01T16:02:52.138Z
worker-1 info: [schedules/watch.js] calling rulesTest() @ 2016-02-01T16:02:52.139Z
worker-1 info: [schedules/watch.js] rulesTest() exited...
worker-1 info: [schedules/watch.js] taskBackgroundProcess(), err: undefined
worker-1 info: [schedules/watch.js] rulesTest() exited...

```

Notice how rulesTest() is called not on the 10 second boundary, and is called back to back immediately..  taskBackgroundProcess() is called once and never again.

**Comment by pgkehle** [comment]
@santigimeno that code you sent over is working fine directly, and in PM2, and with using the updated lodash instead of underscore (well, worked with both). 

I'll check to see if there is something in the code that is breaking it.  But in theory, right, it wouldn't matter what the contents are.

**Comment by santigimeno** [comment]
@pgkehle your last example works fine in my computer. What `node-schedule` version are you running? What nodejs version?

**Comment by pgkehle** [comment]
nodejs v5.5.0, node-schedule v0.6.0

**Comment by slavafomin** [comment]
I'm also using latest versions of both Node.js and the library, i.e. `5.5.0` and `0.6.0`.

**Comment by santigimeno** [comment]
Ok that's it! It looks like another manifestation of https://github.com/node-schedule/node-schedule/issues/225. Just try the code from https://github.com/node-schedule/node-schedule/tree/v1.0-pre and let me know how it goes

**Comment by pgkehle** [comment]
That looks to have it.  

Updated code:

```
    var ruleCfgs = [];
    ruleCfgs.push({name: 'rulesTest', cron: '*/10 * * * * *'});
    ruleCfgs.push({name: 'taskBackgroundProcess', cron: '5,15,25,35,45,55 * * * * *'});

    _.each(ruleCfgs, function (ruleCfg) {

      log.info('ruleCfg.cron: ' + JSON.stringify(ruleCfg.cron, null, 4));

      jobs[ruleCfg.name] = nodeScheduler.scheduleJob(ruleCfg.cron, function () {
        log.info('calling ' + ruleCfg.name + '() @ ' + moment().toISOString());
        schedules[ruleCfg.name](function () {
          log.info('exited ' + ruleCfg.name + '()...');
        });
      });
    });

```

output:

```
worker-1 info: [schedules/watch.js] called; starting monitoring processes...
worker-1 info: [schedules/watch.js] calling taskBackgroundProcess() @ 2016-02-01T16:40:35.820Z
worker-1 info: [schedules/watch.js] calling rulesTest() @ 2016-02-01T16:40:40.815Z
worker-1 info: [schedules/watch.js] calling taskBackgroundProcess() @ 2016-02-01T16:40:45.818Z
worker-1 info: [schedules/watch.js] calling rulesTest() @ 2016-02-01T16:40:50.815Z


```

**Comment by ebr** [comment]
Very interestingly, this started happening to us only yesterday (or in the last couple of days, at most).
Changing from cron-like syntax to recurrence rules solved it for us, but it's so odd that the issue would crop up simultaneously for at least 2 users (@slavafomin & myself) on or around this date.

I was using very straight cron rules (run at 9 AM every day: `0 9 0 0 0`), without specifying the seconds, as they were optional as per documentation. I believe specifying the seconds also fixed it in at least one case, but I had to scramble and refactor to recurrence rules to get things working, so didn't get much time to test. 

FYI I'm using node 0.12.9 across the board.

**Comment by ebr** [comment]
hah, ok, not weird at all, looking at #225 . amazing.

**Comment by schulzetenberg** [comment]
Moving to V1.0-pre fixed the issue for me.

`npm install https://github.com/node-schedule/node-schedule/tarball/v1.0-pre`

**Comment by santigimeno** [comment]
Fixed in https://github.com/node-schedule/node-schedule/commit/2468fb7ea49791d3cb19c4c8eec43abf3c04c793 and released as version `1.0.0`.
Major props to @albertdatui for the excellent fix.

**Comment by slavafomin** [comment]
Thank you guys!
